docs/proposals/20200511-clusterctl-extensible-template-processing.md

```diff
@@ -0,0 +1,344 @@
+
+![Figure 1](./images/clusterctl-extensible-templates/pkgCalls.png)
+
+<div align="center">Package Method Calls for `clusterctl config cluster`</div>
+
+
+The `TemplateClient` will contain a field adhering to the interface - `YamlProcessor`.
+
+![Figure 2](./images/clusterctl-extensible-templates/templateClient.png)
+
+![Figure 3](./images/clusterctl-extensible-templates/yamlProcessor.png)
+
+The `YamlProcessor` will provide three behaviors.
+
+1. `ArtifactName` will return the name of the template artifacts that need to be retrieved from the source.
+
+   - For the default `YamlProcessor`, it will be `cluster-template.yaml`. If the
+     flavor option is provided, it will be `cluster-template-<flavor>.yaml`.
+   - For the ytt `YamlProcessor`, it could be something that follows a separate
+     convention such as
+     `cluster-template-<template-engine-type>.<compression-format>`. This should
+     include the template files, library files, starlark assertion files and
+     default values files. Regarding the flavors specification, we could follow
+     the convention [specified in this example][ytt-multiple-envs-example].
+
+   Once we retrieve the artifact blob of bytes we can then perform the required
+   two functions.
+
+2. `GetVariables` is responsible for parsing the template artifact blob of bytes and
+   providing a list of variables that the template requires.
+
+3. `Process` is responsible for parsing the artifact blob of bytes and will
+   return the final manifest with values retrieved from the `VariablesGetter`.
+
+The flag (`--experimental-template-engine`) value can be read in when
+instantiating the `TemplateClient` and the appropriate type can be created.
+This follows a similar pattern to the `RepositoryClient` and `Repository`
+types.
+
+
+#### Notes
+- Access to using the specific template engines will be available only as part
+  of an experimental flag.
+
+- The yaml processor libraries will be located in a `contrib` subfolder. For
+  example, it could be located within `cluster-api/cmd/clusterctl`.
+
+- Another method of defining a yaml processor instead of using its library
+  would be to shell out to the templating engine binary/executable.
+  The advantage of this is that the templating engine is not a dependency within
+  the clusterctl codebase.
+  The disadvantage is that the user needs to ensure the binary is installed
+  separately which disrupts the UX of clusterctl. If each infrastructure
+  provider provides different template styles, the user would need to install
+  multiple templating tool binaries.
+### Risks and Mitigations
+- Support path for supporting extended templating tools such as ytt within
+  clusterctl. That is, if there are templating issues, who owns them and how
+  do we address them. This is for each templating engine that is included in
+  clusterctl.  
+  However since we’ve decided to put extended templating
+  libraries within a **contrib** folder, it will be the responsibility of the
```

Comment:
In order to properly address this, IMO we should update our contributor's guide to rule contribs (this can be done in a separate PR).

ideally, contribs can be assimilated to experiments in terms of graduation/deprecation rules for ensuring that the new addition to the codebase is maintained at least up to its maturity (see https://cluster-api.sigs.k8s.io/contributing#experiments ).

Reply:
@fabriziopandini @vincepri Should we create a separate story for including guidelines for **contrib** folder in our `CONTRIBUTING.md`?

Reply:
I created this issue https://github.com/kubernetes-sigs/cluster-api/issues/3074 to track this work.

Reply:
The `contrib` should really be the current `exp` folder. W.r.t to the text above, I don't think we've defined what's an experiment/contrib lifecycle for clusterctl.